lib/mongo/socket/ssl.rb

```diff
@@ -116,30 +118,43 @@ def connectable?
 
       def create_context(options)
         context = OpenSSL::SSL::SSLContext.new
-        set_cert(context, options) if options[:ssl_cert]
-        set_key(context, options) if options[:ssl_key]
+        set_cert(context, options)
+        set_key(context, options)
         set_cert_verification(context, options) unless options[:ssl_verify] == false
         context
       end
 
       def set_cert(context, options)
-        context.cert = OpenSSL::X509::Certificate.new(File.open(options[:ssl_cert]))
+        if options[:ssl_cert]
+          context.cert = OpenSSL::X509::Certificate.new(File.open(options[:ssl_cert]))
+        elsif options[:ssl_cert_string]
+          context.cert = OpenSSL::X509::Certificate.new(options[:ssl_cert_string])
+        elsif options[:ssl_cert_object]
+          context.cert = options[:ssl_cert_object]
+        end
       end
 
       def set_key(context, options)
-        if options[:ssl_key_pass_phrase]
-          context.key = OpenSSL::PKey::RSA.new(File.open(options[:ssl_key]),
-                                               options[:ssl_key_pass_phrase])
-        else
-          context.key = OpenSSL::PKey::RSA.new(File.open(options[:ssl_key]))
+        passphrase = options[:ssl_key_pass_phrase]
+        if options[:ssl_key]
+          context.key = passphrase ? OpenSSL::PKey.read(File.open(options[:ssl_key]), passphrase) : OpenSSL::PKey.read(File.open(options[:ssl_key]))
+        elsif options[:ssl_key_string]
+          context.key = passphrase ? OpenSSL::PKey.read(options[:ssl_key_string], passphrase) : OpenSSL::PKey.read(options[:ssl_key_string])
+        elsif options[:ssl_key_object]
+          context.key = options[:ssl_key_object]
```

Comment:
We delegate the checking of the correct type of object to the underlying SSL socket. If we want we can introduce type checks also here